encoder/vaapiencoder_hevc.cpp

```diff
@@ -0,0 +1,1515 @@
+        bit_writer_put_bits_uint32(bitwriter, 1, 1);
+
+        /* reserved zero 44bits */
+        bit_writer_put_bits_uint32(bitwriter, 0, 12);
+        bit_writer_put_bits_uint32(bitwriter, 0, 16);
+        bit_writer_put_bits_uint32(bitwriter, 0, 16);
+    }
+
+    /* general_level_idc */
+    bit_writer_put_bits_uint32(bitwriter, vps_general_level_idc, 8);
+
+    if (max_num_sub_layers) {
+        /* Max layer is zero. and sub layer info is not present
+        * if it can support multi layers, the sub layer info should be present.
+        * TBD
+        */
+    }
+}
+
+
+class VaapiEncStreamHeaderHEVC
+{
+    typedef std::vector<uint8_t> Header;
+public:
+
+    VaapiEncStreamHeaderHEVC();
+    VaapiEncStreamHeaderHEVC(VaapiEncoderHEVC* encoder) {m_encoder = encoder;};
+
+    void setVPS(const VAEncSequenceParameterBufferHEVC* const sequence, VaapiProfile profile)
+    {
+        ASSERT(m_vps.empty());
+        BitWriter bs;
+        bit_writer_init (&bs, 128 * 8);
+        bit_writer_write_vps (&bs, sequence, profile);
+        bsToHeader(m_vps, bs);
+        bit_writer_clear (&bs, TRUE);
+    }
+
+    void setSPS(const VAEncSequenceParameterBufferHEVC* const sequence, VaapiProfile profile)
+    {
+        ASSERT(m_vps.size() && m_sps.empty());
+        BitWriter bs;
+        bit_writer_init (&bs, 128 * 8);
+        bit_writer_write_sps (&bs, sequence, profile);
+        bsToHeader(m_sps, bs);
+        bit_writer_clear (&bs, TRUE);
+    }
+
+    void addPPS(const VAEncPictureParameterBufferHEVC* const picParam)
+    {
+        ASSERT(m_sps.size() && m_pps.empty());
+        BitWriter bs;
+        bit_writer_init (&bs, 128 * 8);
+        bit_writer_write_pps (&bs, picParam);
+        bsToHeader(m_pps, bs);
+        bit_writer_clear (&bs, TRUE);
+    }
+
+    void generateCodecConfig()
+    {
+        std::vector<Header*> headers;
+
+        ASSERT(m_vps.size() && m_sps.size() && (m_sps.size() > 4)&& m_pps.size() && m_headers.empty());
+
+        headers.push_back(&m_vps);
+        headers.push_back(&m_sps);
+        headers.push_back(&m_pps);
+        uint8_t sync[] = {0, 0, 0, 1};
+        for (int i = 0; i < headers.size(); i++) {
+            m_headers.insert(m_headers.end(), sync, sync + N_ELEMENTS(sync));
+            appendHeaderWithEmulation(*headers[i]);
+        }
+    }
+
+    Encode_Status getCodecConfig(VideoEncOutputBuffer *outBuffer)
+    {
+        ASSERT((outBuffer && outBuffer->format == OUTPUT_CODEC_DATA) || outBuffer->format == OUTPUT_EVERYTHING);
+        if (outBuffer->bufferSize < m_headers.size())
+            return ENCODE_BUFFER_TOO_SMALL;
+        if (m_headers.empty())
+            return ENCODE_NO_REQUEST_DATA;
+        std::copy(m_headers.begin(), m_headers.end(), outBuffer->data);
+        outBuffer->dataSize = m_headers.size();
+        outBuffer->flag |= ENCODE_BUFFERFLAG_CODECCONFIG;
+        return ENCODE_SUCCESS;
+    }
+private:
+    BOOL bit_writer_write_vps (
```

Comment:
put bit_writer_write_vps here because I want to keep it same as bit_writer_write_sps and bit_writer_write_pps.  bit_writer_write_sps  will read VaapiEncoderHEVC's members.